File: examples/governance/compliance_domain/lib/compliance_domain/governance_policy/governance_policy.rb
require 'hecks/model'

module ComplianceDomain
  class GovernancePolicy
    autoload :Requirement, "compliance_domain/governance_policy/requirement"
    autoload :Lifecycle, "compliance_domain/governance_policy/lifecycle"

    include Hecks::Model

    attribute :name
    attribute :description
    attribute :category
    attribute :framework_id
    attribute :effective_date
    attribute :review_date
    attribute :requirements, default: [], freeze: true
    attribute :status

    # State predicates — see lifecycle.rb for full state machine
    def draft?; status == "draft"; end
    def active?; status == "active"; end
    def suspended?; status == "suspended"; end
    def retired?; status == "retired"; end

    VALID_CATEGORY = ["regulatory", "internal", "ethical", "operational"].freeze unless defined?(VALID_CATEGORY)

    private

    def validate!
      raise ValidationError, "name can't be blank" if name.nil? || (name.respond_to?(:empty?) && name.empty?)
      raise ValidationError, "category can't be blank" if category.nil? || (category.respond_to?(:empty?) && category.empty?)
      if category && !VALID_CATEGORY.include?(category)
        raise ValidationError, "category must be one of: #{VALID_CATEGORY.join(', ')}, got: #{category}"
      end
    end
  end
end

File: examples/governance_static_go/domain/governance_policy.go
package domain

import (
	"time"
	"github.com/google/uuid"
)

type GovernancePolicy struct {
	ID        string    `json:"id"`
	Name string `json:"name"`
	Description string `json:"description"`
	Category string `json:"category"`
	FrameworkId string `json:"framework_id"`
	EffectiveDate time.Time `json:"effective_date"`
	ReviewDate time.Time `json:"review_date"`
	Requirements []Requirement `json:"requirements"`
	Status string `json:"status"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func NewGovernancePolicy(name string, description string, category string, frameworkId string, effectiveDate time.Time, reviewDate time.Time, requirements []Requirement, status string) *GovernancePolicy {
	a := &GovernancePolicy{
		ID:        uuid.New().String(),
		Name: name,
		Description: description,
		Category: category,
		FrameworkId: frameworkId,
		EffectiveDate: effectiveDate,
		ReviewDate: reviewDate,
		Requirements: requirements,
		Status: status,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *GovernancePolicy) Validate() error {
	if a.Name == "" {
		return &ValidationError{Field: "name", Message: "name can't be blank"}
	}
	if a.Category == "" {
		return &ValidationError{Field: "category", Message: "category can't be blank"}
	}
	return nil
}
